[SecuritySolution] Decouple timeline SO types from server representation types

janmonschke opened the pull request:
## Summary

https://github.com/elastic/security-team/issues/6479

In this PR we're separating the types for the `Timeline` saved object and the type of it's API response equivalent. Doing so will allow us to make changes to either representations individually which is necessary for versioning the SO and routes (individually) in the future.

The saved object definition of timeline and its API equivalent now live in `*/saved_object.ts` and `*/api.ts` respectively. A clean cut of these two, now distinct types. You will encounter a lot of duplication of types in these files which is unavoidable. In the future, depending on how both representations evolve when versioned, these two definitions will diverge.

You will notice that only few types (and values) defined in `saved_object.ts` are exported. They are only used for the conversion logic. They are not exported so they're not accidentally required by frontend or server code that is not dealing with the conversion. The exported types all start with `SavedObject*` to clearly mark them as SO representations.

The conversion files (`convert_saved_object_to_savedtimeline.ts` and `security_solution/server/lib/timeline/saved_object/timelines/index.ts`) have been updated to use the new representations and there is no implicit conversion between them (e.g. through spreading or rest parameters). In some places, an explicit conversion of fields was necessary (e.g. to translate between timeline types).

The bulk of the changes are updates of `import` statements to change the import to `**/api.ts`. If you are on a security team other than the investigations team, you're most likely only required to look at those import changes :)


### Checklist


- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

janmonschke committed:
decouple timeline SO types from server representation types

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

janmonschke committed:
decouple note so types from server representation types

janmonschke committed:
fix broken/missing type imports

janmonschke committed:
remove rejected Promise

michaelolo24 commented on x-pack/plugins/security_solution/common/types/timeline/server.ts:
```diff
@@ -0,0 +1,57 @@
+import type { NoteResult } from './note';
+import type { PinnedEvent } from './pinned_event';
+
+export interface TimelineServerResponse {
```
Thanks for separating this out

janmonschke replied:
Haha, yeah. I think the type will eventually live in that file but for now it's not used.

janmonschke replied:
It's all in `api.ts` now (https://github.com/elastic/kibana/pull/158566/commits/16c7a5fa83d26288cd508bd3cf6314bb8f94b717 and following commits)

janmonschke committed:
move more type defs to `api`

janmonschke committed:
fix timline type imports

janmonschke committed:
fix duplicate import

janmonschke committed:
remove timeline template types from general type file

janmonschke committed:
rename saved object types to SavedObject*

janmonschke committed:
correctly type stored SO to server type

janmonschke committed:
move most of the saved object types to saved_object.ts

Just missing `SavedObjectTimelineType` which is a beast, since it's used everywhere

janmonschke committed:
SavedObjectTimelineType -> TimelineType

janmonschke committed:
remove TimelineStatus from index.ts and update imports

janmonschke committed:
move `RowRendererId` to api.ts

janmonschke committed:
fix more type errors

janmonschke committed:
fix types

janmonschke committed:
fix some imports

janmonschke committed:
rename types in `api.ts` back to original names

janmonschke committed:
fix one more import

janmonschke committed:
fix more types

janmonschke committed:
make sure SO conversion is using the correct types

janmonschke committed:
Migrate away from duplicated types of `@kbn/timelines-plugin`

The ones we were not able to migrate away from immediately were marked as `deprecated`

janmonschke committed:
use new CellValueElementProps and RowRenderer types

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

janmonschke committed:
Revert "[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'"

This reverts commit 2fde8816cd929eb548e038feef31cf13b0eff89f.

janmonschke committed:
Revert "use new CellValueElementProps and RowRenderer types"

This reverts commit cf91d2b1156cae959f9995cbd7cbca3a41130f6a.

janmonschke committed:
mark more types as deprecated

michaelolo24 commented:
### Files not noted here have codeowners of `elastic/security-solution`, but not listed here due to file count

**Specific Codeowners:**

### elastic/actionable-observability

* x-pack/plugins/observability/public/components/alerts_table/render_cell_value.test.tsx
* x-pack/plugins/observability/public/components/alerts_table/render_cell_value.tsx

### elastic/protections-experience

* x-pack/plugins/security_solution/public/threat_intelligence/use_investigate_in_timeline.ts

### elastic/security-defend-workflows

* x-pack/test/security_solution_endpoint/apps/endpoint/endpoint_solution_integrations.ts
* x-pack/test/security_solution_endpoint/apps/endpoint/responder.ts

### elastic/security-detection-engine

* x-pack/plugins/security_solution/public/common/components/sourcerer/index.test.tsx
* x-pack/plugins/security_solution/public/common/components/sourcerer/temporary.tsx
* x-pack/plugins/security_solution/public/common/components/sourcerer/utils.tsx

### elastic/security-detection-rule-management

* x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/install_prebuilt_rules_and_timelines/install_prebuilt_rules_and_timelines_route.ts

### elastic/security-detections-response

* x-pack/plugins/security_solution/cypress/objects/timeline.ts

### elastic/security-threat-hunting-explore

* x-pack/plugins/security_solution/public/common/lib/cell_actions/add_to_timeline.tsx
* x-pack/plugins/security_solution/public/overview/components/detection_response/hooks/use_navigate_to_timeline.tsx
* x-pack/plugins/security_solution/public/overview/components/recent_timelines/index.tsx
* x-pack/plugins/security_solution/public/overview/components/recent_timelines/recent_timelines.tsx

janmonschke committed:
Merge branch 'main' into decouple-timeline-so-types-from-server-representation-types

janmonschke committed:
fix type error

janmonschke committed:
revert SO server logic changes

janmonschke committed:
Merge branch 'main' into decouple-timeline-so-types-from-server-representation-types

maryam-saeidi reviewed (approved):
AO changes LGTM

stephmilovic reviewed (approved):
LGTM for explore, thanks!

yctercero reviewed (approved):
Codeowners for security-detection-engine lgtm.

maximpn reviewed (approved):
Rule Management LGTM
